### DISCUSSION TITLE
Godot Jolt is now available for download!

### DISCUSSION THREAD

**Discussion by mihe** [jrouwe/JoltPhysics#544] [Show and tell]
(The cat's already out of the bag, but I figured I'd make it official. 🙂)

I've just published version [0.1.0-stable](https://github.com/godot-jolt/godot-jolt/releases/tag/v0.1.0-stable) of my project [Godot Jolt](https://github.com/godot-jolt/godot-jolt), which is an extension for the [Godot game engine](https://godotengine.org/) that integrates Jolt as a (mostly) drop-in replacement for Godot's custom physics engine.

While the project is far from finished, and there likely will be many more bugs uncovered, this is still a major milestone for me, as it marks the end of the first phase (and months of work), where I've tried to implement as much of Godot's default physics API as possible.

Coverage didn't exactly end up being 100%, and is probably closer to something like 90%, but I'm still very pleased (and frankly impressed) with how well Jolt and Godot have meshed together.

I also want to take this opportunity to say thank you @jrouwe for being patient, professional and shockingly responsive in our interactions, despite being under no obligation to be either one.

While this isn't exactly Horizon Forbidden West, here's a clip of me playing a [slightly modified](https://github.com/mihe/robo-blast) version of GDQuest's demo game RoboBlast, while using Godot Jolt:

https://github.com/godotengine/godot-proposals/assets/4884246/c218fd38-f9aa-49fb-af87-210dab3a2c68

**Comment by jrouwe** [comment]
Hey, congratulations with this milestone! I've enjoyed working with you too! It was a lot more work than I thought it would be, more than 1/2 year and you certainly haven't been slacking! Looks like there is also quite a bit of interest (200+ stars already).

What are the most important things in the missing 10%? I've been looking through your issue list and I think I could come up with a function that converts from the traditional softness/damping properties to Jolt's frequency/damping properties. Is there more?

**Reply by mihe** [reply]
> It was a lot more work than I thought it would be

You and me both! I figured it would be in the order of months, but I've run over even my pessimistic estimation by about two months. I severely underestimated just how much work making an actual functioning release would be.

> What are the most important things in the missing 10%?

Off the top like this I would say the softness/damping stuff with the constraints is definitely the biggest pain point. People seem to be left wanting with regards to the stability of constraints in Godot Physics, and start to seek out alternative physics implementations, so not being able to accommodate for that right now is unfortunate.

The plan is to expose custom constraint nodes that better map to Jolt's frequency/damping, but there are some major gotchas there right now. Specifically, the C# scripting doesn't support custom nodes exposed by extensions, so you would never be able to interact with them procedurally with anything but GDScript. So supporting the default constraints and all their properties would be a major boon.

I suspect some people will also get tripped up on the difference in how collision layers/masks behave compared to Godot Physics. The behavior in Godot Jolt is more akin to Godot 3, and changing that behavior was highly requested. I'm not sure how to best summarize it, but here's [the original proposal](https://github.com/godotengine/godot-proposals/issues/2775) for the change. I've described it as allowing for "one-way collisions" in the documentation, but I'm not sure that makes much sense. The culprit on my side would be [this code](https://github.com/godot-jolt/godot-jolt/blob/9239130817b8d419313b765fb10638569b66b811/src/spaces/jolt_layer_mapper.cpp#L174-L192). I've experimented with just not doing the two-way check there, but ran into issues that I can't exactly recall right now.

I'm sure I'll think of more stuff in the morning. I'll keep you posted. :)

**Reply by mihe** [reply]
Custom integrators is another thing I've had people ask about, but I must admit I have no sense of how popular of a feature it actually is though.

Internally this is called `omit_force_integration`, which might be a more telling name. It's essentially just telling the physics engine to [stop integrating](https://github.com/godotengine/godot/blob/a67d37f7cffe7f31c68e971280950d487ea99e2c/servers/physics_3d/godot_body_3d.cpp#L628-L651) gravity and any explicitly applied forces, for situations where you want to manipulate the velocities yourself I guess. Any collision response would still be integrated though. You can then also manually run the [default force integration](https://github.com/godotengine/godot/blob/a67d37f7cffe7f31c68e971280950d487ea99e2c/servers/physics_server_3d.cpp#L55-L79) if you need to.

While I could probably implement this on my side to some degree, it would complicate things a bit, and there are some edge-cases that I can't cover for. For example, someone applying a bunch of forces to a body, then setting `custom_integrator` to `true` right after, and still ending up with those forces being applied by Jolt at the next step.

(Note that I'm currently integrating gravity [on my side](https://github.com/godot-jolt/godot-jolt/blob/9239130817b8d419313b765fb10638569b66b811/src/objects/jolt_body_impl_3d.cpp#L654-L680), due to needing things like point gravity.)

**Reply by jrouwe** [reply]
Started a discussion about the masks here: https://github.com/godot-jolt/godot-jolt/issues/400

Added a comment for custom integrators here: https://github.com/godot-jolt/godot-jolt/issues/332

And I will think about this one: https://github.com/godot-jolt/godot-jolt/issues/331

And hopefully I can provide something for converting stiffness/damping to frequency/damping

**Comment by jankrassnigg** [comment]
Congratulations! That's a huge achievement :)
